### PR TITLE
add zipped export result csv

### PR DESCRIPTION
**Pull request by ghazwarhili**

### CODE REVIEW

File: src/test/java/org/gridsuite/shortcircuit/server/ShortCircuitAnalysisControllerTest.java [hunk 1]
```diff
@@ -381,6 +396,36 @@ public void runTest() throws Exception {
...
 
+            // export zipped csv result
+            result = mockMvc.perform(post(
+                            "/" + VERSION + "/results/{resultUuid}/csv", RESULT_UUID)
+                            .contentType(MediaType.APPLICATION_JSON)
+                            .content(CSV_HEADERS))
+                    .andExpectAll(status().isOk(), content().contentType(MediaType.APPLICATION_OCTET_STREAM))
+                    .andReturn();
+            byte[] zippedResult = result.getResponse().getContentAsByteArray();
+            try (ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(zippedResult);
```
Comment (thangqp) [inline src/test/java/org/gridsuite/shortcircuit/server/ShortCircuitAnalysisControllerTest.java:424]: Should extract below unzip code as we did in sensitivity to make test code more clean,
see https://github.com/gridsuite/sensitivity-analysis-server/pull/60/files#diff-24b2d5e07dcacf5c505fffd71b1d39006c833f6b41ccf1cfa5e0cb8b151839faR29

File: src/main/java/org/gridsuite/shortcircuit/server/ShortCircuitException.java [hunk 2]
```diff
@@ -17,7 +17,10 @@
 public class ShortCircuitException extends RuntimeException {
 
     public enum Type {
-        BUS_OUT_OF_VOLTAGE
+        BUS_OUT_OF_VOLTAGE,
+        RESULT_NOT_FOUND,
+        INVALID_EXPORT_PARAMS,
```
Comment (thangqp) [inline src/main/java/org/gridsuite/shortcircuit/server/ShortCircuitException.java:22]: INVALID_EXPORT_PARAMS is not yet used!

File: src/main/java/org/gridsuite/shortcircuit/server/ShortCircuitException.java [hunk 3]
```diff
@@ -17,7 +17,10 @@
 public class ShortCircuitException extends RuntimeException {
```
Comment (thangqp) [inline src/main/java/org/gridsuite/shortcircuit/server/ShortCircuitException.java:17]: Must implement RestResponseEntityExceptionHandler to return correct http code for each type of ShortCircuitException, for example 

@ControllerAdvice
public class RestResponseEntityExceptionHandler {
@ExceptionHandler(SensibilityAnalysisException.class)
protected ResponseEntity<Object> handleSensibilityAnalysisException(SensibilityAnalysisException exception) {
return switch (exception.getType()) {
case RESULT_NOT_FOUND -> ResponseEntity.status(HttpStatus.NOT_FOUND).body(exception.getMessage());
case FILE_EXPORT_ERROR -> ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(exception.getMessage());
case INVALID_EXPORT_PARAMS -> ResponseEntity.status(HttpStatus.BAD_REQUEST).body(exception.getMessage());
default -> ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
};
}
}

File: src/main/java/org/gridsuite/shortcircuit/server/RestResponseEntityExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
+/**
+ * Copyright (c) 2022, RTE (http://www.rte-france.com)
```
Comment (thangqp) [inline src/main/java/org/gridsuite/shortcircuit/server/RestResponseEntityExceptionHandler.java:2]: Copyright (c) 2024,

Comment (ghazwarhili) [reply src/main/java/org/gridsuite/shortcircuit/server/RestResponseEntityExceptionHandler.java]: done

File: src/main/java/org/gridsuite/shortcircuit/server/ShortCircuitException.java (hunk 3 above)
Comment (ghazwarhili) [reply src/main/java/org/gridsuite/shortcircuit/server/ShortCircuitException.java:17]: done

File: src/main/java/org/gridsuite/shortcircuit/server/ShortCircuitException.java (hunk 2 above)
Comment (ghazwarhili) [reply src/main/java/org/gridsuite/shortcircuit/server/ShortCircuitException.java:22]: I used it

File: src/main/java/org/gridsuite/shortcircuit/server/service/ShortCircuitService.java [hunk 5]
```diff
@@ -110,6 +119,80 @@ private static ShortCircuitAnalysisResultEntity sortByElementId(ShortCircuitAnal
...
+            throw new ShortCircuitException(FILE_EXPORT_ERROR, e.getMessage());
+        }
+    }
+
+    public byte[] getZippedCsvExportResult(UUID resultUuid, CsvTranslation csvTranslation) {
+        ShortCircuitAnalysisResult result = getResult(resultUuid, FaultResultsMode.FULL);
+        if (result == null) {
+            throw new ShortCircuitException(RESULT_NOT_FOUND, "The short circuit analysis result '" + resultUuid + "' does not exist");
+        }
+        List<String> headersList = csvTranslation.headersCsv();
```
[the remark this answers is not in the document]
Comment (ghazwarhili) [reply src/main/java/org/gridsuite/shortcircuit/server/service/ShortCircuitService.java:193]: done

File: src/test/java/org/gridsuite/shortcircuit/server/ShortCircuitAnalysisControllerTest.java (hunk 1 above)
Comment (ghazwarhili) [reply src/test/java/org/gridsuite/shortcircuit/server/ShortCircuitAnalysisControllerTest.java]: done

File: src/main/java/org/gridsuite/shortcircuit/server/service/ShortCircuitService.java [hunk 6]
```diff
@@ -110,6 +116,78 @@ private static ShortCircuitAnalysisResultEntity sortByElementId(ShortCircuitAnal
...
+                    }
+                } else {
+                    csvWriter.writeRow(List.of("", "", "", ""));
+                }
+            }
+
+            csvWriter.close();
+            return outputStream.toByteArray();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
```
[the remark this answers is not in the document]
Comment (ghazwarhili) [reply src/main/java/org/gridsuite/shortcircuit/server/service/ShortCircuitService.java]: done

File: src/main/java/org/gridsuite/shortcircuit/server/service/ShortCircuitService.java [hunk 7]
```diff
@@ -110,6 +116,78 @@ private static ShortCircuitAnalysisResultEntity sortByElementId(ShortCircuitAnal
         return result;
     }
 
+    public byte[] getZippedCsvExportResult(UUID resultUuid, String headersCsv) {
+        ShortCircuitAnalysisResult result = getResult(resultUuid, FaultResultsMode.FULL);
+
+        if (result == null) {
+            return new byte[0];
```
[the remark this answers is not in the document]
Comment (ghazwarhili) [reply src/main/java/org/gridsuite/shortcircuit/server/service/ShortCircuitService.java]: done

File: src/main/java/org/gridsuite/shortcircuit/server/ShortCircuitController.java [hunk 8]
```diff
@@ -82,6 +82,18 @@ public ResponseEntity<ShortCircuitAnalysisResult> getResult(@Parameter(descripti
                 : ResponseEntity.notFound().build();
     }
 
+    @PostMapping(value = "/results/{resultUuid}/csv", produces = APPLICATION_OCTET_STREAM_VALUE)
+    @Operation(summary = "Get a short circuit analysis csv result from the database")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The short circuit analysis csv export"),
+        @ApiResponse(responseCode = "404", description = "Short circuit analysis result has not been found")})
+    public ResponseEntity<byte[]> getZippedCsvExportFaultResult(
+            @Parameter(description = "Result UUID") @PathVariable("resultUuid") UUID resultUuid,
+            @RequestBody String headersCsv) {
```
[the remark this answers is not in the document]
Comment (ghazwarhili) [reply src/main/java/org/gridsuite/shortcircuit/server/ShortCircuitController.java]: done

File: src/main/java/org/gridsuite/shortcircuit/server/dto/CsvTranslation.java [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package org.gridsuite.shortcircuit.server.dto;
```
Comment (thangqp) [inline src/main/java/org/gridsuite/shortcircuit/server/dto/CsvTranslation.java:1]: copy right

File: src/main/java/org/gridsuite/shortcircuit/server/dto/CsvTranslation.java [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+import lombok.*;
+
+import java.util.List;
+import java.util.Map;
+
```
Comment (thangqp) [inline src/main/java/org/gridsuite/shortcircuit/server/dto/CsvTranslation.java:7]: author

**Review by thangqp — APPROVED** [review]

File: src/test/java/org/gridsuite/shortcircuit/server/ShortCircuitAnalysisControllerTest.java [hunk 11]
```diff
@@ -189,9 +218,9 @@ private static void assertResultsEquals(ShortCircuitAnalysisResult result, org.g
 
     private static void assertPagedFaultResultsEquals(ShortCircuitAnalysisResult result, List<org.gridsuite.shortcircuit.server.dto.FaultResult> faultResults) {
         assertEquals(result.getFaultResults().size(), faultResults.size());
-        List<FaultResult> orderedFaultResults = result.getFaultResults().stream().sorted(Comparator.comparing(fr -> fr.getFault().getId())).collect(Collectors.toList());
+        var orderedFaultResults = faultResults.stream().sorted(Comparator.comparing(fr -> fr.getFault().getId())).collect(Collectors.toList());
```
Comment (anistouri) [inline src/test/java/org/gridsuite/shortcircuit/server/ShortCircuitAnalysisControllerTest.java:221]: Avoid using 'var'

File: src/main/java/org/gridsuite/shortcircuit/server/dto/CsvTranslation.java [hunk 12]
```diff
@@ -0,0 +1,12 @@
+package org.gridsuite.shortcircuit.server.dto;
+
+import lombok.*;
```
Comment (anistouri) [inline src/main/java/org/gridsuite/shortcircuit/server/dto/CsvTranslation.java:3]:
```suggestion
import lombok.Builder;
```

Comment (ghazwarhili) [reply src/main/java/org/gridsuite/shortcircuit/server/dto/CsvTranslation.java]: done

**Review by anistouri — APPROVED** [review]

**Pull request merged**